File: gateway/test/gateway/permissions_test.exs
defmodule Gateway.PermissionsTest do
  use ExUnit.Case, async: true

  alias Gateway.Permissions

  defp test_vectors_path do
    [
      "/app/testvectors/permissions_vectors.json",
      "/testvectors/permissions_vectors.json",
      Path.expand("../../../testvectors/permissions_vectors.json", __DIR__),
      Path.expand("../../testvectors/permissions_vectors.json", __DIR__),
      Path.expand("testvectors/permissions_vectors.json", File.cwd!())
    ]
    |> Enum.find(&File.exists?/1)
  end

  describe "Cross-language Permission Engine - Elixir Golden Parity (#60)" do
    test "loads and verifies all 45 golden test vectors" do
      path = test_vectors_path()
      assert path != nil, "testvectors/permissions_vectors.json not found"

      vectors =
        path
        |> File.read!()
        |> Jason.decode!()

      assert length(vectors) >= 40

      Enum.each(vectors, fn tc ->
        actual =
          Permissions.resolve_channel(
            tc["guild_id"],
            tc["owner_id"],
            tc["user_id"],
            tc["roles"],
            tc["overwrites"]
          )

        expected = tc["expected"]

        assert actual == expected,
               "Vector failed: #{tc["name"]} (#{tc["description"]})\nexpected: #{expected}, got: #{actual}"
      end)
    end

    test "resolve_guild handles owner, admin, and role unions" do
      assert Permissions.resolve_guild(100, 999, 999, []) == Permissions.all_permissions()

      assert Permissions.resolve_guild(100, 999, 42, [%{"id" => 1, "permissions" => Permissions.administrator()}]) ==
               Permissions.all_permissions()

      assert Permissions.resolve_guild(100, 999, 42, [
               %{"id" => 1, "permissions" => Permissions.view_channel()},
               %{"id" => 2, "permissions" => Permissions.send_messages()}
             ]) == Bitwise.bor(Permissions.view_channel(), Permissions.send_messages())
    end

    test "can? checks permissions correctly" do
      perms = Bitwise.bor(Permissions.view_channel(), Permissions.send_messages())

      assert Permissions.can?(perms, Permissions.view_channel())
      assert Permissions.can?(perms, Permissions.send_messages())
      refute Permissions.can?(perms, Permissions.administrator())
      refute Permissions.can?(perms, Permissions.ban_members())
    end

    test "verifies 29 canonical permission constants and all_permissions" do
      constants = [
        {"create_instant_invite", Permissions.create_instant_invite(), 0},
        {"kick_members", Permissions.kick_members(), 1},
        {"ban_members", Permissions.ban_members(), 2},
        {"administrator", Permissions.administrator(), 3},
        {"manage_channels", Permissions.manage_channels(), 4},
        {"manage_guild", Permissions.manage_guild(), 5},
        {"add_reactions", Permissions.add_reactions(), 6},
        {"view_audit_log", Permissions.view_audit_log(), 7},
        {"priority_speaker", Permissions.priority_speaker(), 8},
        {"stream", Permissions.stream(), 9},
        {"view_channel", Permissions.view_channel(), 10},
        {"send_messages", Permissions.send_messages(), 11},
        {"send_tts_messages", Permissions.send_tts_messages(), 12},
        {"manage_messages", Permissions.manage_messages(), 13},
        {"embed_links", Permissions.embed_links(), 14},
        {"attach_files", Permissions.attach_files(), 15},
        {"read_message_history", Permissions.read_message_history(), 16},
        {"mention_everyone", Permissions.mention_everyone(), 17},
        {"use_external_emojis", Permissions.use_external_emojis(), 18},
        {"view_guild_insights", Permissions.view_guild_insights(), 19},
        {"connect", Permissions.connect(), 20},
        {"speak", Permissions.speak(), 21},
        {"mute_members", Permissions.mute_members(), 22},
        {"deafen_members", Permissions.deafen_members(), 23},
        {"move_members", Permissions.move_members(), 24},
        {"use_vad", Permissions.use_vad(), 25},
        {"change_nickname", Permissions.change_nickname(), 26},
        {"manage_nicknames", Permissions.manage_nicknames(), 27},
        {"manage_roles", Permissions.manage_roles(), 28}
      ]

      union =
        Enum.reduce(constants, 0, fn {name, val, shift}, acc ->
          expected_val = Bitwise.bsl(1, shift)
          assert val == expected_val, "#{name} has value #{val}, expected #{expected_val}"
          assert Bitwise.band(acc, val) == 0, "#{name} overlaps with previous bits"
          Bitwise.bor(acc, val)
        end)

      assert union == Permissions.all_permissions()
      assert Permissions.all_permissions() == Bitwise.bsl(1, 29) - 1
    end
  end

  describe "can_view?/3" do
    alias Gateway.Guild.Cache

    @guild_id "77000000000000001"
    @channel_id "77000000000000010"
    @owner_id "77000000000000099"
    @member_id "77000000000000100"
    @role_id "77000000000000200"

    setup do
      Cache.put_guild(%{
        "id" => @guild_id,
        "name" => "Perms Test Guild",
        "owner_id" => @owner_id,
        "channels" => [%{"id" => @channel_id, "name" => "secret-room"}]
      })

      Cache.put_guild_roles(@guild_id, [])
      Cache.put_member_roles(@member_id, @guild_id, [])
      Cache.put_channel_overwrites(@channel_id, [])

      :ok
    end

    test "guild owner can always view channel" do
      assert Permissions.can_view?(@owner_id, @channel_id, @guild_id)
    end

    test "non-member cannot view channel" do
      refute Permissions.can_view?("random_user_999", @channel_id, @guild_id)
    end

    test "member with @everyone VIEW_CHANNEL can view channel" do
      Cache.put_member_roles(@member_id, @guild_id, [])
      Cache.put_guild_roles(@guild_id, [
        %{"id" => @guild_id, "name" => "@everyone", "position" => 0, "permissions" => Permissions.view_channel()}
      ])

      assert Permissions.can_view?(@member_id, @channel_id, @guild_id)
    end

    test "member with role VIEW_CHANNEL can view channel even if @everyone denies it" do
      Cache.put_member_roles(@member_id, @guild_id, [@role_id])
      Cache.put_guild_roles(@guild_id, [
        %{"id" => @guild_id, "name" => "@everyone", "position" => 0, "permissions" => 0},
        %{"id" => @role_id, "name" => "VIP", "position" => 1, "permissions" => Permissions.view_channel()}
      ])

      assert Permissions.can_view?(@member_id, @channel_id, @guild_id)
    end

    test "channel overwrite denying VIEW_CHANNEL blocks member" do
      Cache.put_member_roles(@member_id, @guild_id, [@role_id])
      Cache.put_guild_roles(@guild_id, [
        %{"id" => @role_id, "name" => "VIP", "position" => 1, "permissions" => Permissions.view_channel()}
      ])
      Cache.put_channel_overwrites(@channel_id, [
        %{"target_id" => @member_id, "target_type" => 1, "allow" => 0, "deny" => Permissions.view_channel()}
      ])

      refute Permissions.can_view?(@member_id, @channel_id, @guild_id)
    end
  end
end

File: client/src/lib/permissions.test.ts
import { describe, expect, it } from 'vitest'
import vectors from '../../../testvectors/permissions_vectors.json'
import * as Perms from './permissions'

describe('Cross-language Permission Engine - TypeScript Golden Parity (#60)', () => {
  it('loads and verifies all 45 golden test vectors', () => {
    expect(vectors.length).toBeGreaterThanOrEqual(40)

    for (const tc of vectors) {
      const actual = Perms.resolveChannelPermissions(
        tc.guild_id,
        tc.owner_id,
        tc.user_id,
        tc.roles,
        tc.overwrites
      )
      const expected = BigInt(tc.expected)

      expect(
        actual,
        `Vector failed: ${tc.name} (${tc.description})`
      ).toBe(expected)
    }
  })

  it('resolveGuildPermissions handles owner, admin, and role unions', () => {
    // Owner bypass
    expect(Perms.resolveGuildPermissions('100', '999', '999', [])).toBe(Perms.ALL_PERMISSIONS)

    // Admin bypass in role
    expect(
      Perms.resolveGuildPermissions('100', '999', '42', [
        { id: '1', permissions: Perms.ADMINISTRATOR.toString() },
      ])
    ).toBe(Perms.ALL_PERMISSIONS)

    // Normal union
    expect(
      Perms.resolveGuildPermissions('100', '999', '42', [
        { id: '1', permissions: Perms.VIEW_CHANNEL.toString() },
        { id: '2', permissions: Perms.SEND_MESSAGES.toString() },
      ])
    ).toBe(Perms.VIEW_CHANNEL | Perms.SEND_MESSAGES)
  })

  it('hasPermission evaluates single and combined permissions', () => {
    const perms = Perms.VIEW_CHANNEL | Perms.SEND_MESSAGES | Perms.ATTACH_FILES

    expect(Perms.hasPermission(perms, Perms.VIEW_CHANNEL)).toBe(true)
    expect(Perms.hasPermission(perms, Perms.SEND_MESSAGES)).toBe(true)
    expect(Perms.hasPermission(perms, Perms.ATTACH_FILES)).toBe(true)
    expect(Perms.hasPermission(perms, Perms.VIEW_CHANNEL | Perms.SEND_MESSAGES)).toBe(true)
    expect(Perms.hasPermission(perms, Perms.BAN_MEMBERS)).toBe(false)
    expect(Perms.hasPermission(perms, Perms.ADMINISTRATOR)).toBe(false)
    expect(Perms.hasPermission(perms, Perms.VIEW_CHANNEL | Perms.BAN_MEMBERS)).toBe(false)
  })

  it('verifies 29 canonical permission constants and ALL_PERMISSIONS', () => {
    const bits: Array<[string, bigint, bigint]> = [
      ['CREATE_INSTANT_INVITE', Perms.CREATE_INSTANT_INVITE, 0n],
      ['KICK_MEMBERS', Perms.KICK_MEMBERS, 1n],
      ['BAN_MEMBERS', Perms.BAN_MEMBERS, 2n],
      ['ADMINISTRATOR', Perms.ADMINISTRATOR, 3n],
      ['MANAGE_CHANNELS', Perms.MANAGE_CHANNELS, 4n],
      ['MANAGE_GUILD', Perms.MANAGE_GUILD, 5n],
      ['ADD_REACTIONS', Perms.ADD_REACTIONS, 6n],
      ['VIEW_AUDIT_LOG', Perms.VIEW_AUDIT_LOG, 7n],
      ['PRIORITY_SPEAKER', Perms.PRIORITY_SPEAKER, 8n],
      ['STREAM', Perms.STREAM, 9n],
      ['VIEW_CHANNEL', Perms.VIEW_CHANNEL, 10n],
      ['SEND_MESSAGES', Perms.SEND_MESSAGES, 11n],
      ['SEND_TTS_MESSAGES', Perms.SEND_TTS_MESSAGES, 12n],
      ['MANAGE_MESSAGES', Perms.MANAGE_MESSAGES, 13n],
      ['EMBED_LINKS', Perms.EMBED_LINKS, 14n],
      ['ATTACH_FILES', Perms.ATTACH_FILES, 15n],
      ['READ_MESSAGE_HISTORY', Perms.READ_MESSAGE_HISTORY, 16n],
      ['MENTION_EVERYONE', Perms.MENTION_EVERYONE, 17n],
      ['USE_EXTERNAL_EMOJIS', Perms.USE_EXTERNAL_EMOJIS, 18n],
      ['VIEW_GUILD_INSIGHTS', Perms.VIEW_GUILD_INSIGHTS, 19n],
      ['CONNECT', Perms.CONNECT, 20n],
      ['SPEAK', Perms.SPEAK, 21n],
      ['MUTE_MEMBERS', Perms.MUTE_MEMBERS, 22n],
      ['DEAFEN_MEMBERS', Perms.DEAFEN_MEMBERS, 23n],
      ['MOVE_MEMBERS', Perms.MOVE_MEMBERS, 24n],
      ['USE_VAD', Perms.USE_VAD, 25n],
      ['CHANGE_NICKNAME', Perms.CHANGE_NICKNAME, 26n],
      ['MANAGE_NICKNAMES', Perms.MANAGE_NICKNAMES, 27n],
      ['MANAGE_ROLES', Perms.MANAGE_ROLES, 28n],
    ]

    let union = 0n
    for (const [name, val, shift] of bits) {
      expect(val).toBe(1n << shift)
      expect(union & val, `Overlap detected for ${name}`).toBe(0n)
      union |= val
    }

    expect(union).toBe(Perms.ALL_PERMISSIONS)
    expect(Perms.ALL_PERMISSIONS).toBe((1n << 29n) - 1n)
  })

  it('identifies private channels where @everyone is denied VIEW_CHANNEL', () => {
    const guildId = '100'

    const publicChannelOverwrites: Perms.OverwriteLike[] = []
    const isPublic = publicChannelOverwrites.some(
      (ow) => Number(ow.target_type) === 0 && String(ow.target_id) === guildId && Perms.hasPermission(ow.deny, Perms.VIEW_CHANNEL)
    )
    expect(isPublic).toBe(false)

    const privateChannelOverwrites: Perms.OverwriteLike[] = [
      { target_id: guildId, target_type: 0, allow: '0', deny: Perms.VIEW_CHANNEL.toString() },
      { target_id: '200', target_type: 0, allow: Perms.VIEW_CHANNEL.toString(), deny: '0' },
    ]
    const isPrivate = privateChannelOverwrites.some(
      (ow) => Number(ow.target_type) === 0 && String(ow.target_id) === guildId && Perms.hasPermission(ow.deny, Perms.VIEW_CHANNEL)
    )
    expect(isPrivate).toBe(true)
  })

  it('validates MANAGE_CHANNELS permission gating', () => {
    expect(Perms.hasPermission(0n, Perms.MANAGE_CHANNELS)).toBe(false)
    expect(Perms.hasPermission(Perms.MANAGE_CHANNELS, Perms.MANAGE_CHANNELS)).toBe(true)
    expect(Perms.hasPermission(Perms.ADMINISTRATOR, Perms.ADMINISTRATOR)).toBe(true)
  })

  it('resolves SEND_MESSAGES gating for message input', () => {
    const guildId = '100'
    const ownerId = '999'
    const memberId = '42'

    const baseRoles: Perms.RoleLike[] = [
      { id: guildId, permissions: (Perms.VIEW_CHANNEL | Perms.SEND_MESSAGES).toString() },
    ]

    // 1. Normal channel allows send
    const normalPerms = Perms.resolveChannelPermissions(guildId, ownerId, memberId, baseRoles, [])
    expect(Perms.hasPermission(normalPerms, Perms.SEND_MESSAGES)).toBe(true)

    // 2. Read-only channel with SEND_MESSAGES denied for @everyone
    const readOnlyOverwrites: Perms.OverwriteLike[] = [
      { target_id: guildId, type: 0, allow: '0', deny: Perms.SEND_MESSAGES.toString() },
    ]
    const readOnlyPerms = Perms.resolveChannelPermissions(guildId, ownerId, memberId, baseRoles, readOnlyOverwrites)
    expect(Perms.hasPermission(readOnlyPerms, Perms.SEND_MESSAGES)).toBe(false)

    // 3. Member with VIP role allowed SEND_MESSAGES overrides @everyone deny
    const vipRoles: Perms.RoleLike[] = [
      ...baseRoles,
      { id: '200', permissions: '0' },
    ]
    const vipOverwrites: Perms.OverwriteLike[] = [
      { target_id: guildId, type: 0, allow: '0', deny: Perms.SEND_MESSAGES.toString() },
      { target_id: '200', type: 0, allow: Perms.SEND_MESSAGES.toString(), deny: '0' },
    ]
    const vipPerms = Perms.resolveChannelPermissions(guildId, ownerId, memberId, vipRoles, vipOverwrites)
    expect(Perms.hasPermission(vipPerms, Perms.SEND_MESSAGES)).toBe(true)

    // 4. Owner bypasses read-only deny unconditionally
    const ownerPerms = Perms.resolveChannelPermissions(guildId, ownerId, ownerId, [], readOnlyOverwrites)
    expect(Perms.hasPermission(ownerPerms, Perms.SEND_MESSAGES)).toBe(true)

    // 5. Server base role override: @everyone has SEND_MESSAGES=0, but custom role has SEND_MESSAGES=1
    const baseRolesWithoutEveryoneSend: Perms.RoleLike[] = [
      { id: guildId, permissions: Perms.VIEW_CHANNEL.toString() }, // @everyone: no SEND_MESSAGES
      { id: '300', permissions: Perms.SEND_MESSAGES.toString() },   // @nerds: has SEND_MESSAGES
    ]
    const resolvedMemberWithRole = Perms.resolveChannelPermissions(
      guildId,
      ownerId,
      memberId,
      baseRolesWithoutEveryoneSend,
      []
    )
    expect(Perms.hasPermission(resolvedMemberWithRole, Perms.SEND_MESSAGES)).toBe(true)

    // Member without the custom role cannot send messages
    const resolvedMemberWithoutRole = Perms.resolveChannelPermissions(
      guildId,
      ownerId,
      memberId,
      [{ id: guildId, permissions: Perms.VIEW_CHANNEL.toString() }],
      []
    )
    expect(Perms.hasPermission(resolvedMemberWithoutRole, Perms.SEND_MESSAGES)).toBe(false)
  })
})
